eventio/simtel/objects.py

```diff
@@ -665,27 +683,19 @@ def parse_data_field(self):
         n_pixels = read_from('<i', self)[0]
         n_gains = read_from('<h', self)[0]
 
-        if raw['data_red_mode'] == 2:
-            offset_hg8 = read_from('<h', self)[0]
-            scale_hg8 = read_from('<h', self)[0]
-            if scale_hg8 <= 0:
-                scale_hg8 = 1
-
-        if raw['zero_sup_mode'] == 0.:
-            if raw['data_red_mode'] == 0:
-                # before version 3, it was just uint16
-                raw['adc_sums'] = np.array([
-                    read_vector_of_uint32_scount_differential(self, n_pixels)
-                    for gain in range(n_gains)
-
-                ])
-        if 'adc_sums' not in raw:
+        if raw['data_red_mode'] != 0 or raw['zero_sup_mode'] != 0:
             raise NotImplementedError(
                 'Currently no support for data_red_mode {} or zero_sup_mode{}'.format(
                     raw['data_red_mode'], raw['zero_sup_mode'],
                 )
             )
 
+        raw['adc_sums'] = []
+        data = self.read()
+        raw['adc_sums'], bytes_read = unsigned_varint_arrays_differential(
+            data, n_arrays=n_gains, n_elements=n_pixels
+        )
```

Comment:
I am always a little concerned if our parsers really parse the entire payload. Since if they do not parse the entire payload, this is an early hint for a bug. 

In this case we could `assert nearly_equal(bytes_read, len(data))` ... 

On the other hand, as soon as we have an event_source() up and running we can compare the contents with pyhessio and there is no need to rely on such hints ... 

Also, I would rather assert this only in tests and not in production code. But for tests to do this assertion, we would have to return bytes_read .. which is also not perfect.

Do you have any ideas here? 
At the moment however .. any `self.read()` invalidates the part of the tests, which tries to make sure the complete payload is really consumed by the parser.

Reply:
Yes, I wanted to do that, but for the current dataset it the assertion was wrong, although the result was the same as what pyhessio read